### DISCUSSION TITLE
Added: LDM Notes for Nov. 13, 2019

### DISCUSSION THREAD

**Discussion by agocke** [dotnet/csharplang#2968] [General]
https://github.com/dotnet/csharplang/blob/master/meetings/2019/LDM-2019-11-13.md

Agenda:

1. Discriminated unions
2. Improved analysis of `[MaybeNull]T`

**Comment by alrz** [comment]
> ```cs
> enum class Expr
> {
>     enum class BinaryOperator
>     {
>         Add(Expr left, Expr right),
>         Multiply(Expr left, Expr right)
>     }
> }
> ```
> 

Instead of this, I would prefer a way to define DUs as sibling classes for more complex type hierarchies, Roslyn AST is a good example here.

In general, for both records and DUs, I think there could be a **compact** and **expanded** syntax forms, the latter would provide more control over things. The IDE could help to switch between the two.

**Comment by erik-kallen** [comment]
Since you are considering improving nullable reference type handling, can you please have a look at #2830 and the problem it describes as well. I don't think nullable reference types are very useful in web apps until the problem described in that issue is fixed.

**Comment by MgSam** [comment]
Could someone post a full example using `enum class`? I'm not fully understanding the semantics from the proposal without a full example declaration and usage. Also, it seems to require records as a pre-requisite- is that the case?

**Comment by 333fred** [comment]
The examples in the proposal are about as fleshed-out as we have, currently. We haven't yet talked about how you would construct these, for example. And yes, record types are a prerequisite for unions.
